pkg/collector/remoteconfig/remote_config.go

```diff
@@ -0,0 +1,107 @@
+	"github.com/DataDog/datadog-agent/pkg/collector"
+	"github.com/DataDog/datadog-agent/pkg/remoteconfig/state"
+	"github.com/DataDog/datadog-agent/pkg/secrets"
+	pkglog "github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// Scheduler is the structure used to receive remote-config checks, and add or remove
+// them in the agent's scheduler
+type Scheduler struct {
+	scheduler     *collector.CheckScheduler
+	runningChecks []integration.Config
+}
+
+type agentIntegration struct {
+	Name       string            `json:"name"`
+	Instances  []json.RawMessage `json:"instances"`
+	InitConfig json.RawMessage   `json:"init_config"`
+}
+
+// secretsDecrypt allows tests to intercept calls to secrets.Decrypt.
+var secretsDecrypt = secrets.Decrypt
+
+// NewScheduler creates an instance of a remote config integration scheduler
+func NewScheduler(scheduler *collector.CheckScheduler) *Scheduler {
+	return &Scheduler{
+		runningChecks: make([]integration.Config, 0),
+		scheduler:     scheduler,
+	}
+}
+
+// IntegrationScheduleCallback is called at every AGENT_INTEGRATIONS to schedule/unschedule integrations
+func (sc *Scheduler) IntegrationScheduleCallback(updates map[string]state.RawConfig, applyStateCallback func(string, state.ApplyStatus)) {
+	if sc.scheduler == nil {
+		pkglog.Debugf("Scheduler is not ready to start remote-config integrations")
+		return
+	}
+
+	// Unschedule every integrations, even if they haven't changed
+	sc.scheduler.Unschedule(sc.runningChecks)
+	sc.runningChecks = make([]integration.Config, 0)
+
+	// Now schedule everything
+	for cfgPath, intg := range updates {
+		var d agentIntegration
+		err := json.Unmarshal(intg.Config, &d)
+		if err != nil {
+			pkglog.Errorf("Can't decode agent configuration provided by remote-config: %v", err)
+		}
+
+		shouldSchedule := true
+		applyStateCallback(cfgPath, state.ApplyStatus{State: state.ApplyStateUnacknowledged})
+
+		configToSchedule := integration.Config{
+			Name:       d.Name,
+			Instances:  []integration.Data{},
+			InitConfig: integration.Data(d.InitConfig),
+		}
+		for _, instance := range d.Instances {
+			// Resolve the ENC[] configuration, and fetch the actual secret in the config backend
+			decryptedInstance, err := secretsDecrypt(instance, d.Name)
+			if err != nil {
+				pkglog.Errorf("Couldn't decrypt remote-config integration %s secret: %s", d.Name, err)
+				applyStateCallback(cfgPath, state.ApplyStatus{
+					State: state.ApplyStateError,
+					Error: err.Error(),
+				})
+				shouldSchedule = false
+				continue
+			}
+			configToSchedule.Instances = append(configToSchedule.Instances, integration.Data(decryptedInstance))
+		}
+
+		// We either apply all the instances of the config or none, no partial apply
+		if !shouldSchedule {
+			break
+		}
+
+		// Schedule all the checks from this configuration
+		err = sc.scheduler.ScheduleWithErrors(configToSchedule)
+		pkglog.Infof("Scheduled %d instances of %s check with remote-config", len(d.Instances), d.Name)
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ Question:** ‏In case of errors, won't it be confusing to have a log that says "scheduled" then "there were error(s)"?